pkg/plugin/system/system_test.go

```diff
@@ -0,0 +1,252 @@
+					return fmt.Errorf("test error")
+				})
+			})
+		})
+
+		g.Expect(e).To(MatchError("test error"))
+	})
+
+	t.Run("parent interceptor errors override child errors", func(t *testing.T) {
+		g := NewGomegaWithT(t)
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		// Minimal streams
+		var stdout strings.Builder
+		streams := ioutils.Streams{Stdout: &stdout, Stderr: &stdout}
+
+		// empty ctx + args + command
+		ctx := context.Background()
+		args := []string{}
+		cmd := &cobra.Command{
+			Use: "HookOrderTest",
+		}
+		// Required flags for interceptor hooks
+		cmd.PersistentFlags().Bool(rootFlags.InteractiveFlagName, false, "")
+
+		plugin := plugin_mock.NewMockPlugin(ctrl)
+
+		// Expect the callbacks in reverse-order of execution
+		gomock.InOrder(
+			plugin.EXPECT().PostRunHook(gomock.Any(), gomock.Any()),
+			plugin.EXPECT().PostTestHook(gomock.Any(), gomock.Any()),
+			plugin.EXPECT().PostBuildHook(gomock.Any(), gomock.Any()),
+		)
+
+		ps := NewPluginSystem().(*pluginSystem)
+		ps.Configure(streams)
+		ps.plugins.insert(plugin)
+
+		// Hook interceptors
+		buildInterceptor := ps.BuildHooksInterceptor(streams)
+		testInterceptor := ps.TestHooksInterceptor(streams)
+		runInterceptor := ps.RunHooksInterceptor(streams)
+
+		e := buildInterceptor(ctx, cmd, args, func(ctx context.Context, cmd *cobra.Command, args []string) error {
+			return testInterceptor(ctx, cmd, args, func(ctx context.Context, cmd *cobra.Command, args []string) error {
+				runInterceptor(ctx, cmd, args, func(ctx context.Context, cmd *cobra.Command, args []string) error {
+					return fmt.Errorf("error 1")
+				})
+				return fmt.Errorf("error 2")
+			})
+		})
+
+		g.Expect(e).To(MatchError("error 2"))
+	})
+
+	t.Run("should set ExitCode=1 on error of type aspecterrors.ExitError", func(t *testing.T) {
+		g := NewGomegaWithT(t)
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		// Minimal streams
+		var stdout strings.Builder
+		streams := ioutils.Streams{Stdout: &stdout, Stderr: &stdout}
+
+		// empty ctx + args + command
+		ctx := context.Background()
+		args := []string{}
+		cmd := &cobra.Command{
+			Use: "HookOrderTest",
+		}
+		// Required flags for interceptor hooks
+		cmd.PersistentFlags().Bool(rootFlags.InteractiveFlagName, false, "")
+
+		plugin := plugin_mock.NewMockPlugin(ctrl)
+		plugin.EXPECT().PostRunHook(gomock.Any(), gomock.Any())
+
+		ps := NewPluginSystem().(*pluginSystem)
+		ps.Configure(streams)
+		ps.plugins.insert(plugin)
+
+		// Hook interceptors
+		runInterceptor := ps.RunHooksInterceptor(streams)
+		e := runInterceptor(ctx, cmd, args, func(ctx context.Context, cmd *cobra.Command, args []string) error {
+			return &aspecterrors.ExitError{
+				Err:      fmt.Errorf("aspect error"),
+				ExitCode: 123,
+			}
+		})
+
+		g.Expect(e.(*aspecterrors.ExitError).Err).To(MatchError("aspect error"))
+		g.Expect(e.(*aspecterrors.ExitError).ExitCode).To(Equal(1))
```

Comment:
This is failing... does that mean https://github.com/aspect-build/aspect-cli/blob/2cc2cb276b936077943fd30f3f6aa6fa46acb069/pkg/plugin/system/system.go#L190-L195 is not working? Or do you see an issue with my test?

Reply:
It is. Your test needs to actually add a `DoAndReturn` to the mocked method:

```
diff --git a/pkg/plugin/sdk/v1alpha2/plugin/mock/BUILD.bazel b/pkg/plugin/sdk/v1alpha2/plugin/mock/BUILD.bazel
index b99dfa6..facdfbb 100644
--- a/pkg/plugin/sdk/v1alpha2/plugin/mock/BUILD.bazel
+++ b/pkg/plugin/sdk/v1alpha2/plugin/mock/BUILD.bazel
@@ -22,6 +22,7 @@ go_library(
     visibility = ["//visibility:public"],
     deps = [
         "//bazel/buildeventstream/proto",  # keep
+        "//pkg/ioutils",  # keep
         "//pkg/plugin/sdk/v1alpha2/plugin",  # keep
         "@com_github_golang_mock//gomock",  # keep
     ],
diff --git a/pkg/plugin/system/system_test.go b/pkg/plugin/system/system_test.go
index 1c721d6..ac8ddfa 100644
--- a/pkg/plugin/system/system_test.go
+++ b/pkg/plugin/system/system_test.go
@@ -231,7 +231,17 @@ func TestPluginSystem(t *testing.T) {
 		cmd.PersistentFlags().Bool(rootFlags.InteractiveFlagName, false, "")
 
 		plugin := plugin_mock.NewMockPlugin(ctrl)
-		plugin.EXPECT().PostRunHook(gomock.Any(), gomock.Any())
+		plugin.EXPECT().
+			PostRunHook(gomock.Any(), gomock.Any()).
+			DoAndReturn(func(
+				isInteractiveMode bool,
+				promptRunner ioutils.PromptRunner,
+			) interface{} {
+				return &aspecterrors.ExitError{
+					Err:      fmt.Errorf("plugin error"),
+					ExitCode: 123,
+				}
+			})
 
 		ps := NewPluginSystem().(*pluginSystem)
 		ps.Configure(streams)
```

Reply:
Oh, I'm just being stupid and returning the error from the wrong spot 🤦‍♂️ - thank you! lol

I guess there's 2 types of errors then (and I was only producing the second): the plugin invocations and the "next" invocation (which might be more plugin invocations)